File: examples/webhook-receiver/node/receiver.js
/**
 * RepoRoller Webhook Receiver — Node.js Example
 * ===============================================
 * Demonstrates how to receive and verify RepoRoller outbound webhook notifications.
 *
 * Requirements:
 *   node >= 18  (built-in `crypto` module; no extra dependencies)
 *
 * Usage:
 *   WEBHOOK_SECRET="your-shared-secret-value" node receiver.js
 *
 * The server listens on port 8080 and accepts POST /webhook requests.
 *
 * See docs/notifications.md for full webhook documentation.
 */

'use strict';

const crypto = require('crypto');
const http = require('http');

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET;
if (!WEBHOOK_SECRET) {
    console.error('ERROR: WEBHOOK_SECRET environment variable is not set');
    process.exit(1);
}

const PORT = parseInt(process.env.PORT || '8080', 10);

// ---------------------------------------------------------------------------
// Signature verification
// ---------------------------------------------------------------------------

/**
 * Returns true if the given signature header matches the HMAC-SHA256 of the body.
 *
 * Uses `crypto.timingSafeEqual` to prevent timing attacks.
 *
 * @param {Buffer} rawBody - The raw, unparsed request body bytes.
 * @param {string} signatureHeader - Value of `X-RepoRoller-Signature-256` header.
 *                                    Expected format: `sha256=<hex>`.
 * @returns {boolean}
 */
function verifySignature(rawBody, signatureHeader) {
    if (!signatureHeader || !signatureHeader.startsWith('sha256=')) {
        return false;
    }

    const receivedHex = signatureHeader.slice('sha256='.length);

    const hmac = crypto.createHmac('sha256', WEBHOOK_SECRET);
    hmac.update(rawBody);
    const computedHex = hmac.digest('hex');

    // Constant-time comparison — never use === here.
    const receivedBuf = Buffer.from(receivedHex, 'hex');
    const computedBuf = Buffer.from(computedHex, 'hex');

    if (receivedBuf.length !== computedBuf.length) {
        return false;
    }

    return crypto.timingSafeEqual(receivedBuf, computedBuf);
}

// ---------------------------------------------------------------------------
// Event handlers
// ---------------------------------------------------------------------------

/**
 * Handle a `repository.created` event.
 * @param {object} payload - Parsed event payload.
 */
function handleRepositoryCreated(payload) {
    const {
        event_id,
        organization,
        repository_name,
        repository_url,
        created_by,
        visibility,
        template_name,
        team,
        content_strategy,
    } = payload;

    console.info(
        `Repository created — event_id=${event_id} org=${organization} ` +
        `name=${repository_name} url=${repository_url} created_by=${created_by} ` +
        `visibility=${visibility} template=${template_name ?? '(none)'} ` +
        `team=${team ?? '(none)'} strategy=${content_strategy}`
    );

    // Add your integration logic here:
    //   - Post a Slack / Teams notification
    //   - Register the repo in a service catalog
    //   - Trigger a CI provisioning pipeline
    //   - Update an asset inventory database
}

// ---------------------------------------------------------------------------
// HTTP server
// ---------------------------------------------------------------------------

const server = http.createServer((req, res) => {
    const url = new URL(req.url, `http://${req.headers.host}`);

    if (req.method !== 'POST' || url.pathname !== '/webhook') {
        res.writeHead(404);
        res.end('Not Found');
        return;
    }

    // Collect raw body bytes BEFORE parsing — signature covers the raw body.
    const chunks = [];
    req.on('data', (chunk) => chunks.push(chunk));
    req.on('end', () => {
        const rawBody = Buffer.concat(chunks);
        const signatureHeader = req.headers['x-reporoller-signature-256'] || '';

        // 1. Verify signature.
        if (!verifySignature(rawBody, signatureHeader)) {
            console.warn(`Rejected request with invalid signature from ${req.socket.remoteAddress}`);
            res.writeHead(401);
            res.end('Unauthorized');
            return;
        }

        // 2. Parse payload.
        let payload;
        try {
            payload = JSON.parse(rawBody.toString('utf8'));
        } catch (err) {
            console.error(`Failed to parse JSON body: ${err.message}`);
            res.writeHead(400);
            res.end('Bad Request');
            return;
        }

        // 3. Dispatch on event type.
        const eventType = payload.event_type;
        if (eventType === 'repository.created') {
            handleRepositoryCreated(payload);
        } else {
            console.info(`Ignoring unknown event type: ${eventType}`);
        }

        // Always acknowledge promptly — processing is fire-and-forget from sender's perspective.
        res.writeHead(204);
        res.end();
    });

    req.on('error', (err) => {
        console.error(`Request error: ${err.message}`);
        res.writeHead(500);
        res.end('Internal Server Error');
    });
});

server.listen(PORT, '0.0.0.0', () => {
    console.info(`RepoRoller webhook receiver listening on port ${PORT}`);
    // In production, terminate TLS at a reverse proxy / load balancer.
    // Your notifications.toml endpoint URL must use https://.
});

File: examples/webhook-receiver/go/receiver.go
// RepoRoller Webhook Receiver — Go Example
// ==========================================
// Demonstrates how to receive and verify RepoRoller outbound webhook notifications.
//
// Requirements:
//
//	Go 1.21+  (only stdlib — no external dependencies)
//
// Usage:
//
//	WEBHOOK_SECRET="your-shared-secret-value" go run receiver.go
//
// The server listens on port 8080 and accepts POST /webhook requests.
//
// See docs/notifications.md for full webhook documentation.
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

var webhookSecret []byte

func init() {
	secret := os.Getenv("WEBHOOK_SECRET")
	if secret == "" {
		slog.Error("WEBHOOK_SECRET environment variable is not set")
		os.Exit(1)
	}
	webhookSecret = []byte(secret)
}

func port() string {
	p := os.Getenv("PORT")
	if p == "" {
		return "8080"
	}
	if _, err := strconv.Atoi(p); err != nil {
		slog.Error("PORT must be a number", "value", p)
		os.Exit(1)
	}
	return p
}

// ---------------------------------------------------------------------------
// Event payload types
// ---------------------------------------------------------------------------

// RepositoryCreatedPayload contains all fields sent in a repository.created event.
// Optional fields use pointer types so absent fields deserialise as nil.
type RepositoryCreatedPayload struct {
	EventType        string            `json:"event_type"`
	EventID          string            `json:"event_id"`
	Timestamp        string            `json:"timestamp"`
	Organization     string            `json:"organization"`
	RepositoryName   string            `json:"repository_name"`
	RepositoryURL    string            `json:"repository_url"`
	RepositoryID     string            `json:"repository_id"`
	CreatedBy        string            `json:"created_by"`
	RepositoryType   *string           `json:"repository_type,omitempty"`
	TemplateName     *string           `json:"template_name,omitempty"`
	ContentStrategy  string            `json:"content_strategy"`
	Visibility       string            `json:"visibility"`
	Team             *string           `json:"team,omitempty"`
	Description      *string           `json:"description,omitempty"`
	CustomProperties map[string]string `json:"custom_properties,omitempty"`
}

// genericPayload is used only to read the event_type field before full deserialisation.
type genericPayload struct {
	EventType string `json:"event_type"`
}

// ---------------------------------------------------------------------------
// Signature verification
// ---------------------------------------------------------------------------

// verifySignature returns true when signatureHeader matches the HMAC-SHA256 of
// rawBody using the shared secret.
//
// Uses hmac.Equal (constant-time) to prevent timing attacks.
func verifySignature(rawBody []byte, signatureHeader string) bool {
	const prefix = "sha256="
	if !strings.HasPrefix(signatureHeader, prefix) {
		return false
	}

	receivedHex := signatureHeader[len(prefix):]
	receivedBytes, err := hex.DecodeString(receivedHex)
	if err != nil {
		return false
	}

	mac := hmac.New(sha256.New, webhookSecret)
	mac.Write(rawBody)
	computed := mac.Sum(nil)

	// Constant-time comparison — never use bytes.Equal here.
	return hmac.Equal(computed, receivedBytes)
}

// ---------------------------------------------------------------------------
// Event handlers
// ---------------------------------------------------------------------------

func handleRepositoryCreated(payload RepositoryCreatedPayload) {
	templateName := "(none)"
	if payload.TemplateName != nil {
		templateName = *payload.TemplateName
	}
	team := "(none)"
	if payload.Team != nil {
		team = *payload.Team
	}

	slog.Info("Repository created",
		"event_id", payload.EventID,
		"org", payload.Organization,
		"name", payload.RepositoryName,
		"url", payload.RepositoryURL,
		"created_by", payload.CreatedBy,
		"visibility", payload.Visibility,
		"template", templateName,
		"team", team,
		"strategy", payload.ContentStrategy,
	)

	// Add your integration logic here:
	//   - Post a Slack / Teams notification
	//   - Register the repo in a service catalog
	//   - Trigger a CI provisioning pipeline
	//   - Update an asset inventory database
}

// ---------------------------------------------------------------------------
// HTTP handler
// ---------------------------------------------------------------------------

func webhookHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
		return
	}

	// Read the full body BEFORE parsing — signature covers the raw bytes.
	rawBody, err := io.ReadAll(io.LimitReader(r.Body, 1<<20)) // 1 MiB limit
	if err != nil {
		slog.Error("Failed to read request body", "error", err)
		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
		return
	}
	defer r.Body.Close()

	// 1. Verify signature.
	sigHeader := r.Header.Get("X-RepoRoller-Signature-256")
	if !verifySignature(rawBody, sigHeader) {
		slog.Warn("Rejected request with invalid signature", "remote_addr", r.RemoteAddr)
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	// 2. Determine event type.
	var env genericPayload
	if err := json.Unmarshal(rawBody, &env); err != nil {
		slog.Error("Failed to parse JSON body", "error", err)
		http.Error(w, "Bad Request", http.StatusBadRequest)
		return
	}

	// 3. Dispatch on event type.
	switch env.EventType {
	case "repository.created":
		var payload RepositoryCreatedPayload
		if err := json.Unmarshal(rawBody, &payload); err != nil {
			slog.Error("Failed to deserialise repository.created payload", "error", err)
			http.Error(w, "Bad Request", http.StatusBadRequest)
			return
		}
		handleRepositoryCreated(payload)

	default:
		slog.Info("Ignoring unknown event type", "event_type", env.EventType)
	}

	// Always acknowledge promptly — processing is fire-and-forget from sender's perspective.
	w.WriteHeader(http.StatusNoContent)
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

func main() {
	addr := fmt.Sprintf("0.0.0.0:%s", port())

	mux := http.NewServeMux()
	mux.HandleFunc("/webhook", webhookHandler)

	slog.Info("RepoRoller webhook receiver listening", "addr", addr)
	// In production, terminate TLS at a reverse proxy / load balancer.
	// Your notifications.toml endpoint URL must use https://.
	if err := http.ListenAndServe(addr, mux); err != nil {
		slog.Error("Server failed", "error", err)
		os.Exit(1)
	}
}
